### ISSUE TITLE
Non-shuffled order sample per-epoch

### ISSUE THREAD

**Issue by xiye17** [Qihoo360/360-LLaMA-Factory#46]
### Reminder

- [x] I have read the README and searched the existing issues.

### System Info

Thanks for the codebase! I am using the codebase to SFT model on a small dataset for 8 epochs. Looking at the loss curve, it seems the orders of samples were not properly shuffled at each epoch. I would like to check if this is   done by design or this is a issue?

![Image](https://github.com/user-attachments/assets/da36e531-b111-4369-a39b-dc94b91c92cf)

### Reproduction

Here are some of the args. I am using sequence parallelism and ds_z3_offload. I am **not** setting packing args
```
    --stage sft \
    --do_train \
    --sequence_parallel_size 4 \
    --lr_scheduler_type linear \
    --warmup_ratio 0.0 \
    --num_train_epochs 8 \
    --deepspeed examples/deepspeed/ds_z3_offload_config.json 
```

**Comment by HaoshengZou** [comment]
It is by design. When using SP, data is shuffled only at the very beginning.

**Comment by xiye17** [comment]
Thank you so much for the prompt response!
Sorry I am not very familiar with Seq Parallel, I wonder if you can briefly explain why we have this design?

**Comment by phamquang-hieu** [comment]
The same issue was raised in the following DeepSpeed's repo: https://github.com/deepspeedai/DeepSpeed/issues/5619
which refers to line 266 in the following implementation of DeepSpeed's engine: https://github.com/deepspeedai/DeepSpeed/blob/master/deepspeed/runtime/pipe/engine.py#L266.

People in [#5619](https://github.com/deepspeedai/DeepSpeed/issues/5619) agreed that setting shuffle to True should not hurt the training process, and there should be a way to make that parameter configurable. Yet it has not been done. 

I assume that is because Seq Parallel aims to train gigantic models with a lot of data, in which case, there is often no second epoch...
